src/kbmod/search/geom.h

```diff
@@ -0,0 +1,226 @@
+    os << rc.to_string();
+    return os;
+}
+
+// A rectangle that also contains it's corner's origin Index with respect to
+// another, larger rectangle.
+struct AnchoredRectangle {
+    Index corner;
+    Index anchor;
+
+    unsigned height;
+    unsigned width;
+
+    const std::string to_string() const {
+        return "AnchoredRectangle(corner:" + corner.to_string() + ", anchor:" + anchor.to_string() +
+               ", height: " + std::to_string(height) + ", width: " + std::to_string(width) + ")";
+    }
+
+    const std::string to_yaml() const {
+        return "{corner: " + corner.to_yaml() + " anchor: " + anchor.to_yaml() +
+               " height: " + std::to_string(height) + " width: " + std::to_string(width) + "}";
+    }
+
+    friend std::ostream& operator<<(std::ostream& os, const AnchoredRectangle& rc);
+};
+
+std::ostream& operator<<(std::ostream& os, const AnchoredRectangle& rc) {
+    os << rc.to_string();
+    return os;
+}
+
+// Given an index component `idx_val`, and a radius `r` around it, returns the
+// start and end index components, and length of the range, that fit in the
+// [0, max_range] limits.
+inline std::tuple<int, int, int> centered_range(int idx_val, const int r, const int max_range) {
+    // pin start to the []0, max_range] range
+    int start = std::max(0, idx_val - r);
+    start = std::min(start, max_range);
+
+    // pin end to the [0, max_range]
+    int end = std::max(0, idx_val + r);
+    end = std::min(max_range, idx_val + r);
+
+    // range is inclusive of the first element, and can not be longer than start
+    // minus max range
+    int length = end - start + 1;
+    length = std::min(length, max_range - start);
+
+    return std::make_tuple(start, end, length);
+}
+
+// get an Eigen block coordinates (top-left, height, width) anchored inside a
+// square matrix of dimensions 2*r+1 (anchor top-left).
+inline AnchoredRectangle anchored_block(const Index& idx, const int r, const unsigned height,
+                                        const unsigned width) {
+    auto [top, bot, rangey] = centered_range(idx.j, r, height);
+    auto [left, right, rangex] = centered_range(idx.i, r, width);
+    assertm(rangey > 0, "Selected block lies outside of the image limits.");
+    assertm(rangex > 0, "Selected block lies outside of the image limits.");
+
+    int anchor_top = std::max(r - idx.j, 0);
+    int anchor_left = std::max(r - idx.i, 0);
+
+    // now it's safe to cast ranges to unsigned
+    return {{top, left}, {anchor_top, anchor_left}, (unsigned)rangey, (unsigned)rangex};
+}
+
+inline auto manhattan_neighbors(const Index& idx, const unsigned width, const unsigned height) {
+    std::array<std::optional<Index>, 4> idxs;
+
+    // top bot
+    if (idx.i >= 0 && idx.i < width) {
+        if (idx.j - 1 >= 0 && idx.j - 1 < height) idxs[0] = {idx.j - 1, idx.i};
+        if (idx.j + 1 >= 0 && idx.j + 1 < height) idxs[1] = {idx.j + 1, idx.i};
```

Comment:
What happens in the case these `if` conditions are not met? Would `idxs[1]` be `nullptr` or some similar indicator?

I think it is worth putting this in a comment. That it returns an array of the 4 neighbors or ??? if the neighbor is out of bounds.

Reply:
it's an optional type so it casts to `false` in C++, and to `None` in Python. I've added a comment.